src/renderer/dx/DxRenderer.cpp

```diff
@@ -229,13 +240,112 @@ _CompileShader(
 #endif
 }
 
+// Routine Description:
+// - Checks if terminal effects are enabled
+// Arguments:
+// Return Value:
+// - True if terminal effects are enabled
+bool DxEngine::_HasTerminalEffects() const noexcept
+{
+    return _terminalEffectsEnabled && (_retroTerminalEffect || _pixelShaderEffect);
+}
+
+// Routine Description:
+// - Disable terminal effects
+// Arguments:
+// Return Value:
+// - Void
+void DxEngine::_DisableTerminalEffects() noexcept
+{
+    _terminalEffectsEnabled = false;
+    _retroTerminalEffect = false;
+    _pixelShaderEffect.reset();
+}
+
+// Routine Description:
+// - Toggles terminal effects off and on. If no terminal effect is configured has no effect
+// Arguments:
+// Return Value:
+// - Void
+void DxEngine::ToggleTerminalEffects()
+{
+    _terminalEffectsEnabled = !_terminalEffectsEnabled;
+    LOG_IF_FAILED(InvalidateAll());
+}
+
+// Routine Description:
+// - Loads pixel shader source depending on _retroTerminalEffect and _pixelShaderEffect
+// Arguments:
+// Return Value:
+// - Pixel shader source code
+std::string DxEngine::_LoadPixelShaderEffect() const
+{
+    try
+    {
+        // If the user specified the legacy option retroTerminalEffect it has precendence
+        if (_retroTerminalEffect)
+        {
+            return std::string{ RETRO_PIXEL_SHADER };
+        }
+
+        // If no pixel shader effect is specified this function shouldn't end up being called.
+        //  If it happens anyway return the error shader
+        if (!_pixelShaderEffect)
+        {
+            return std::string{ ERROR_PIXEL_SHADER };
+        }
+
+        auto pixelShaderEffect = *_pixelShaderEffect;
+
+        const auto pixelShaderPreset = PIXEL_SHADER_PRESETS.find(pixelShaderEffect);
+
+        if (pixelShaderPreset != PIXEL_SHADER_PRESETS.end())
+        {
+            return std::string{ pixelShaderPreset->second };
+        }
+
+        wil::unique_hfile hFile{ CreateFileW(pixelShaderEffect.c_str(),

```

Comment:
TODO: Are there security concerns around this? It's a program that will execute on the GPU. Denial of Service?

Reply:
Might https://docs.microsoft.com/en-us/windows/win32/direct2d/load-a-id2d1image-using-the-filepicker be helpful here?

Reply:
I am not exactly sure how `DirectXTex` fits the purpose here. The pixel shader effect file is text source code that is compiled into a pixel shader. `DirectXTex` and `D3DX11GetImageInfoFromFile` seems more oriented towards loading bitmaps into texture.

Not saying it doesn't apply but from the docs I don't understand how currently.

Reply:
https://github.com/microsoft/Windows-universal-samples/tree/master/Samples/FilePicker/cppwinrt might be more helpful then.

Reply:
That also doesn't _really_ matter. We've already got the path to the file at this point. We don't really need a file picker as a part of the DX renderer.